static/src/javascripts/bootstraps/enhanced/notifications.js

```diff
@@ -0,0 +1,180 @@
+                modules.buttonHandler();
+            });
+        },
+
+        setSubscriptionStatus: function (subscribed) {
+            var subscribeButton = $('.js-notifications-subscribe-link'),
+                subscribeLink = $('.notifications-subscribe-link--follow');
+            isSubscribed = subscribed;
+            if (subscribed) {
+                subscribeLink.addClass('notifications-subscribe-link--has-subscriptions');
+            }
+            subscribeButton[0].textContent = subscribed ?  'Following story' : 'Follow story';
+        },
+
+        buttonHandler: function () {
+            if (isSubscribed) {
+                modules.stopFollowingThis();
+            } else {
+                modules.subscribe();
+            }
+        },
+
+        subscribe: function () {
+            if (modules.subscriptionsEmpty()) {
+                reg.pushManager.subscribe({userVisibleOnly: true}).then(function (pushSubscription) {
+                    sub = pushSubscription;
+                    modules.followThis();
+                });
+            } else {
+                modules.followThis();
+            }
+        },
+
+
+        unSubscribe: function () {
+            if (modules.subscriptionsEmpty()) {
+                sub.unsubscribe().then(function () {
+                }).catch(function (error) {
+                    robust.log('07cm-frontendNotificatons', error);
+                });
+            }
+        },
+
+        subscriptionsEmpty: function () {
+            var subscriptions = userPrefs.get('subscriptions') || [];
+            return subscriptions.length ? false : true;
+        },
+
+        checkSubscriptions: function () {
+            var subscriptions = userPrefs.get('subscriptions') || [];
+            return some(subscriptions, function (sub) {
+                return sub == config.page.pageId;
+            });
+        },
+
+        followThis: function () {
+            var endpoint = '/notification/store';
+
+            modules.updateSubscription(endpoint).then(
+                function () {
+                    var subscriptions = userPrefs.get('subscriptions') || [];
+                    subscriptions.push(config.page.pageId);
+                    userPrefs.set('subscriptions', uniq(subscriptions));
+                    modules.setSubscriptionStatus(true);
+                }
+            );
+        },
+
+        stopFollowingThis: function () {
+
+            var endpoint = '/notification/delete';
+            modules.updateSubscription(endpoint).then(
+                function () {
+                    var subscriptions = userPrefs.get('subscriptions') || [],
+                        newSubscriptions = without(subscriptions, config.page.pageId);
+                    userPrefs.set('subscriptions', uniq(newSubscriptions));
+                    modules.setSubscriptionStatus(false);
+                    if (modules.subscriptionsEmpty()) {
+                        modules.unSubscribe();
```

Comment:
What if the unsubscribe operation fails? The button would still say you're unsubscribed. Perhaps we need to update the state of the button if unsubscribe returns a rejected promise?